# Project Description

OilyGiant a mining company is looking for the best place for a new well. 

Location is chosen based on:
<li> Oil quality and volume of reserve </li>
<li> Volume of reserves in the new wells based on a model</li>
<li> Oil wells with the highest estimated values </li>
<li> Region with the highest total profit for the selected oil wells.</li>

Additional information:
<li> When exploring the region, a study of 500 points is carried with picking the best 200 points for the profit calculation.</li>
<li> The budget for development of 200 oil wells is 100 USD million.</li>
<li> One barrel of raw materials brings 4.5 USD of revenue The revenue from one unit of product is 4,500 dollars (volume of reserves is in thousand barrels).</li>
<li> After the risk evaluation, we are looking to keep only the regions with the risk of losses lower than 2.5%. From the ones that fit the criteria, the region with the highest average profit should be selected. </li>

We have data on oil samples from three regions. The parameters of each oil well in the region are already known. We are looking to build a linear regression model that will help to pick the region with the highest profit margin through profit and risk analysis using the Bootstrapping technique. 

## Import and Review Data

### Importing Libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import recall_score, precision_score, f1_score, roc_auc_score, r2_score, mean_absolute_error, confusion_matrix, accuracy_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelEncoder
from sklearn.datasets import make_hastie_10_2
import matplotlib.pyplot as plt

### Import and Review Data

In [2]:
re0 = pd.read_csv('/datasets/geo_data_0.csv')
display(re0.head(5))
print()
print(re0.shape)
print()
display(re0.describe())
print()
print(re0.duplicated().sum())
print()
print(re0.info())

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647


(100000, 5)



f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347


0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [3]:
re1 = pd.read_csv('/datasets/geo_data_1.csv')
display(re1.head(5))
print()
print(re1.shape)
print()
display(re1.describe())
print()
print(re1.duplicated().sum())
print()
print(re1.info())

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305


(100000, 5)



f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408


0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [4]:
re2 = pd.read_csv('/datasets/geo_data_2.csv')
display(re2.head(5))
print()
print(re2.shape)
print()
display(re2.describe())
print()
print(re2.duplicated().sum())
print()
print(re2.info())

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


(100000, 5)



f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838


0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [5]:
reg0 = re0.drop(columns='id')
reg1 = re1.drop(columns='id')
reg2 = re2.drop(columns='id')

<div class="alert alert-info"> <b> Data Review </b>:
    <li> Each file is for 3 different region's data </li>
    <li> id - unique oil well identifier. This was dropped since it's non-feature data that will not be relevant in training a model. </li>
    <li> f0,f1,f2 - represent different features of the oil wells and are already on same scale. </li>
    <li> product - volume of reserves in the oil well (thousand barrels) </li>
    <li> There are no duplicates or missing values in all files </li>

## Train and test the model for each region

### Split the data
We will be splitting the data into training set and validation set at a ratio of 75:25

In [6]:
features0 = reg0.drop(['product'], axis=1)
target0 = reg0['product']

features0_train, features0_valid, target0_train, target0_valid = train_test_split(
    features0, target0, test_size=0.25, random_state=123)

print(features0_train.shape)
print(features0_valid.shape)

(75000, 3)
(25000, 3)


In [7]:
features1 = reg1.drop(['product'], axis=1)
target1 = reg1['product']

features1_train, features1_valid, target1_train, target1_valid = train_test_split(
    features1, target1, test_size=0.25, random_state=123)

print(features1_train.shape)
print(features1_valid.shape)

(75000, 3)
(25000, 3)


In [8]:
features2 = reg2.drop(['product'], axis=1)
target2 = reg2['product']

features2_train, features2_valid, target2_train, target2_valid = train_test_split(
    features2, target2, test_size=0.25, random_state=123)

print(features2_train.shape)
print(features2_valid.shape)

(75000, 3)
(25000, 3)


### Model training and predictions
Train the model and make predictions for the validation set

#### Region 0

In [9]:
model = LinearRegression()

model.fit(features0_train, target0_train)
predictions_valid0 = model.predict(features0_valid)

#### Region 1

In [10]:
model.fit(features1_train, target1_train)
predictions_valid1 = model.predict(features1_valid)

#### Region 2

In [11]:
model.fit(features2_train, target2_train)
predictions_valid2 = model.predict(features2_valid)

### Predictions and correct answers for validation set
Save the predictions and correct answers for the validation set.

#### Region 0

In [12]:
correct_answer0 = target0_valid.tolist()
predicted_values0 = predictions_valid0.tolist()

val_result0 = pd.DataFrame(
    {'correct_answers0': correct_answer0, 'predictions0': predicted_values0})

#### Region 1

In [13]:
correct_answer1 = target1_valid.tolist()
predicted_values1 = predictions_valid1.tolist()

val_result1 = pd.DataFrame(
    {'correct_answers1': correct_answer1, 'predictions1': predicted_values1})

#### Region 2

In [14]:
correct_answer2 = target2_valid.tolist()
predicted_values2 = predictions_valid2.tolist()

val_result2 = pd.DataFrame(
    {'correct_answers2': correct_answer2, 'predictions2': predicted_values2})

#### Average volume of predicted reserves and model RMSE

### Average volume of predicted reserves

In [15]:
reservmean0pred = val_result0['predictions0'].mean()
reservmean1pred = val_result1['predictions1'].mean()
reservmean2pred = val_result2['predictions2'].mean()

print("Region 0 average volume of predicted reserve in thousand barrels: ", reservmean0pred)
print("Region 1 average volume of predicted reserve in thousand barrels: ", reservmean1pred)
print("Region 2 average volume of predicted reserve in thousand barrels: ", reservmean2pred)

Region 0 average volume of predicted reserve in thousand barrels:  92.54936189116306
Region 1 average volume of predicted reserve in thousand barrels:  69.28001860653976
Region 2 average volume of predicted reserve in thousand barrels:  95.09859933591373


#### Model RMSE

In [16]:
rmse0 = mean_squared_error(target0_valid, predictions_valid0, squared=False)
rmse1 = mean_squared_error(target1_valid, predictions_valid1, squared=False)
rmse2 = mean_squared_error(target2_valid, predictions_valid2, squared=False)

print("Root Mean Squared Error on Validation Set of Region 0:", rmse0)
print("Root Mean Squared Error on Validation Set of Region 1:", rmse1)
print("Root Mean Squared Error on Validation Set of Region 2:", rmse2)

Root Mean Squared Error on Validation Set of Region 0: 37.64786282376176
Root Mean Squared Error on Validation Set of Region 1: 0.8954139804944313
Root Mean Squared Error on Validation Set of Region 2: 40.12803006598514


### Result Analysis

<li> Relatively, Region 2 has the highest mean reserve and Region 1 has the least. </li>
<li> However, the RMSE of the Region 1 is significantly lower than Region 0 or 2, suggesting that current model prediction is the most accurate for Region 1. </li>

## Prep for Gross Profit Calculation

### Key values

#### Volume of Reserves

In [17]:
reg0rsrv = target0
reg1rsrv = target1
reg2rsrv = target2

In [18]:
target_r0 = val_result0['correct_answers0']
target_r1 = val_result1['correct_answers1']
target_r2 = val_result2['correct_answers2']

pred_r0 = val_result0['predictions0']
pred_r1 = val_result1['predictions1']
pred_r2 = val_result2['predictions2']

#### Number of Wells

In [19]:
reg0numwell = re0['id'].nunique()
reg1numwell = re1['id'].nunique()
reg2numwell = re2['id'].nunique()

print('Number of wells in Region 0: ', reg0numwell,
      '\nNumber of wells in Region 1: ', reg1numwell,
      '\nNumber of wells in Region 2: ', reg2numwell)

Number of wells in Region 0:  99990 
Number of wells in Region 1:  99996 
Number of wells in Region 2:  99996


#### Revenue per Barrel

In [20]:
# revenue per barrel
revpb = 4.5*1000
# One barrel of raw materials brings 4.5 USD of revenue.
# The revenue from one unit of product is 4,500 dollars (volume of reserves is in thousand barrels).

# Average potential revenue per region
reg0pr = revpb * reservmean0pred
reg1pr = revpb * reservmean1pred
reg2pr = revpb * reservmean2pred

print('Potential average revenue 0: ', reg0pr,
      '\nPotential average revenue 1: ', reg1pr,
      '\nPotential average revenue 2: ', reg2pr)

Potential average revenue 0:  416472.1285102338 
Potential average revenue 1:  311760.0837294289 
Potential average revenue 2:  427943.6970116118


#### Budget per Well

In [21]:
budg = 100000000
w = 200
budgpw = 100000000/200
print(budgpw)

500000.0


### Volume Needed

Volume of reserves and minimum number of wells sufficient for developing a new well without losses. 

In [22]:
volneed = budgpw/revpb
print('Minimum volume of resesrves needed per well:', volneed)

Minimum volume of resesrves needed per well: 111.11111111111111


In [23]:
break_even_wells = budg/volneed

print("Number of Wells to Break Even: ", break_even_wells)

Number of Wells to Break Even:  900000.0


### Profit steps explained:

<li> Budget per well : \$500,000 </li>
<li> Potential revenue per well = 4500 * number of barrels/product. Based on this and average volume of barrels, and none of them meet the budget per well allocated. <ul>
<li> Potential average revenue 0:  \$416,472 </li>
<li> Potential average revenue 1:  \$311,760 </li>
<li> Potential average revenue 2:  \$427,944 </li></ul></li>
<li> Minimum volume need for a well development is 111.1111 without incurring lost (or to equal budget per well)</li>

## Gross Profit 

### High value wells
Based on predictions

In [24]:
# Sort the DataFrame by 'predictions0' column in descending order
val_result_sorted0 = val_result0.sort_values(
    by='predictions0', ascending=False)
val_result_sorted1 = val_result1.sort_values(
    by='predictions1', ascending=False)
val_result_sorted2 = val_result2.sort_values(
    by='predictions2', ascending=False)

# Select the top wells with the highest predicted values
top_wells_highest_predictions0 = val_result_sorted0.head(200)
top_wells_highest_predictions1 = val_result_sorted1.head(200)
top_wells_highest_predictions2 = val_result_sorted2.head(200)

### Target Volume vs. High Value Wells
We'll compare volume needed to the average of top 200 wells from each region and predict their average reserves.

In [25]:
topm0pred = top_wells_highest_predictions0['predictions0'].sum()
topm1pred = top_wells_highest_predictions1['predictions1'].sum()
topm2pred = top_wells_highest_predictions2['predictions2'].sum()

print("Region 0 average volume of top 200 wells: ", topm0pred)
print("Region 1 average volume of top 200 wells: ", topm1pred)
print("Region 2 average volume of top 200 wells: ", topm2pred)

Region 0 average volume of top 200 wells:  31034.787805093576
Region 1 average volume of top 200 wells:  27755.76198279395
Region 2 average volume of top 200 wells:  29943.443110391767


### Profit 

In [26]:
# pick wells with the highest volume of predictions, then sort prediction values in descending order:
def profit(target, predictions, count):
    top_wells = predictions.sort_values(ascending=False).head(count).index
    total_reserves = target.loc[top_wells].sum()
    revenue = total_reserves * 4500
    profit = revenue - budg
    return profit


reg0profit = profit(target_r0, pred_r0, 200)
reg1profit = profit(target_r1, pred_r1, 200)
reg2profit = profit(target_r2, pred_r2, 200)

print('Profit of Region 0: ', reg0profit)
print('Profit of Region 1: ', reg1profit)
print('Profit of Region 2: ', reg2profit)

Profit of Region 0:  35346709.17261383
Profit of Region 1:  24150866.966815114
Profit of Region 2:  23703438.630213737


### Data Review
<div class="alert alert-info"> Based on prediction and profit, Region 0 would likely to yield highest profit when further developed. Please note that these are estiamtes of the maximum possible profit, as we don't actually have the budget to make the initial measurements in more than 500 locations, and the calculation assumes that we can select from all wells in the region.

## Risk and Profit

### Bootstrapping
We'll use the bootstrapping technique with 1000 samples to find the distribution of profit.

#### Region 0 

In [27]:
state = np.random.RandomState(1234)

profit_subsampling0 = []
for i in range(1000):
    subsample_target0 = target_r0.sample(
        n=500, replace=True, random_state=state)
    subsample_predict0 = pred_r0[subsample_target0.index]
    subsample_profit0 = profit(subsample_target0, subsample_predict0, 200)
    profit_subsampling0.append(subsample_profit0)

#### Region 1

In [28]:
state = np.random.RandomState(2345)
profit_subsampling1 = []
for i in range(1000):
    subsample_target1 = target_r1.sample(
        n=500, replace=True, random_state=state)
    subsample_predict1 = pred_r1[subsample_target1.index]
    subsample_profit1 = profit(subsample_target1, subsample_predict1, 200)
    profit_subsampling1.append(subsample_profit1)

#### Region 2

In [29]:
state = np.random.RandomState(3456)
profit_subsampling2 = []
for i in range(1000):
    subsample_target2 = target_r2.sample(
        n=500, replace=True, random_state=state)
    subsample_predict2 = pred_r2[subsample_target2.index]
    subsample_profit2 = profit(subsample_target2, subsample_predict2, 200)
    profit_subsampling2.append(subsample_profit2)

### Average Profit 
with 95% confidence interval and risk of losses. Loss is negative profit, calculate it as a probability and it'll be expressed as percentage. 

In [30]:
# Calculate average profit
average_profit0 = np.mean(profit_subsampling0)

# Calculate 95% confidence interval
lower_bound0 = np.percentile(profit_subsampling0, 2.5)
upper_bound0 = np.percentile(profit_subsampling0, 97.5)

# Calculate the risk of losses (probability of negative profit)
losses_probability0 = np.mean([profit < 0 for profit in profit_subsampling0])
# Express as a percentage
losses_percentage0 = losses_probability0 * 100

# Display the results
print("Average Profit:", average_profit0)
print("95% Confidence Interval - Lower Bound:", lower_bound0)
print("95% Confidence Interval - Upper Bound:", upper_bound0)
print("Risk of Losses (as a percentage):", losses_percentage0, '%')

Average Profit: 6780221.214204038
95% Confidence Interval - Lower Bound: 806113.8429065847
95% Confidence Interval - Upper Bound: 12773269.099896826
Risk of Losses (as a percentage): 1.0999999999999999 %


In [31]:
# Convert the list to a NumPy array for convenience
profit_subsampling1 = np.array(profit_subsampling1)

# Calculate average profit
average_profit1 = np.mean(profit_subsampling1)

# Calculate 95% confidence interval
lower_bound1 = np.percentile(profit_subsampling1, 2.5)
upper_bound1 = np.percentile(profit_subsampling1, 97.5)

# Calculate the risk of losses (probability of negative profit)
losses_probability1 = np.mean([profit < 0 for profit in profit_subsampling1])
# Express as a percentage
losses_percentage1 = losses_probability1 * 100

# Display the results
print("Average Profit:", average_profit1)
print("95% Confidence Interval - Lower Bound:", lower_bound1)
print("95% Confidence Interval - Upper Bound:", upper_bound1)
print("Risk of Losses (as a percentage):", losses_percentage1, '%')

Average Profit: 6748432.196633914
95% Confidence Interval - Lower Bound: 1851433.7347724729
95% Confidence Interval - Upper Bound: 12340984.646680254
Risk of Losses (as a percentage): 0.2 %


In [32]:
profit_subsampling2 = np.array(profit_subsampling2)
average_profit2 = np.mean(profit_subsampling2)
lower_bound2 = np.percentile(profit_subsampling2, 2.5)
upper_bound2 = np.percentile(profit_subsampling2, 97.5)
losses_probability2 = np.mean([profit < 0 for profit in profit_subsampling2])
losses_percentage2 = losses_probability2 * 100

print("Average Profit:", average_profit2)
print("95% Confidence Interval - Lower Bound:", lower_bound2)
print("95% Confidence Interval - Upper Bound:", upper_bound2)
print("Risk of Losses (as a percentage):", losses_percentage2, '%')

Average Profit: 5561090.588947411
95% Confidence Interval - Lower Bound: -377732.14212882804
95% Confidence Interval - Upper Bound: 12070795.772695841
Risk of Losses (as a percentage): 3.5000000000000004 %


<div class="alert alert-info"> 
    
**Key Findings**:

<li> Profitability: Region 1 exhibits an average profit of approximately \$6,748,432, only slightly below the highest profit region (Region 0). </li>
<li> Risk Assessment: With a mere 0.2% risk of losses, Region 1 demonstrates the lowest probability of negative profit among the three regions. </li> 
<li> Confidence in Profit: The 95% confidence interval for Region 1 remains entirely positive, ensuring a reliable estimation of profit potential. </li>

**Rationalization**:

<li> Balanced Performance: Region 1 strikes a remarkable balance between profitability and risk. Despite marginally lower average profit compared to Region 0, its minimal risk of losses makes it a robust choice. </li>
<li> Stability in Estimates: The confidence intervals imply consistent positive profit estimates, bolstering confidence in the region's viability for development. </li>

**Recommendation**:

Based on these observations, investing in Region 1 for oil well development appears as the most prudent choice. Its balanced performance, dependable profit estimates, and minimal risk make it an optimal selection for maximizing profitability while mitigating potential losses.
